davidverweij/csv2docx
# Type hinting with static type analysis

## Addresses Issue #11

- [ ] ~I added **tests** for changed code.~
- [x] I updated **documentation** for changed code.
- [x] I ran `python run nox -r` successfully



## What Changed (and why)?

  1. Added type annotations to all methods to prevent bugs;
  1. Added mypy package to ``--dev`` and updated `noxfile.py` to do static analysis by default;
  1. Added mypy to pre-commit to validate code before commits, e.g. to PRs;
  1. Made very minor tweak, e.g. removed [uncommented code](https://github.com/davidverweij/csv2docx/commit/1f57615ca24ea3c34ee6dbd5abab8c54126c629c) and bumped package version from `0.0.1` to `0.1.0` as a patch without minor version did not make [semantic sense](http://semver.org/);
  1. Updated documentation to note that we use [static analysis for types on pre-commit](https://github.com/davidverweij/csv2docx/commit/1f57615ca24ea3c34ee6dbd5abab8c54126c629c#diff-04c6e90faac2675aa89e2176d2eec7d8R71).

Running mypy identified a few issues, e.g. (i) reassigning `path` from a `str` to a `path` in `convert(...)` or (ii) the complexity of having the `options` list using multiple types.

## How to test Changes

  1. Pull down changes locally
  1. Run `poetry install && poetry shell`
  1. Run `poetry run nox` and observe that `mypy` runs and is a _success_
  1. Verify that `mypy` pre-commit works: (i) remove a type hint from `convert()`; (ii) add the file to git; and (iii) try to make a commit. This should fail 👍 

Please review @davidverweij and closes #11.
